src/ucp/core/ucp_am.c

```diff
@@ -0,0 +1,700 @@
+                     (worker->am_cbs[hdr->am_id].cb == NULL))) {
+        ucs_warn("UCP Active Message was received with id : %u, but there" 
+                 "is no registered callback for that id", hdr->am_id);
+        return UCS_OK;
+    }
+
+    return ucp_am_handler_common(worker, hdr + 1, sizeof(*hdr),
+                                 am_length - sizeof(*hdr), NULL,
+                                 am_id, UCP_RECV_DESC_FLAG_AM_HDR, 
+                                 am_flags);    
+}
+
+
+
+ucp_am_unfinished_t *
+ucp_am_find_parent(ucp_worker_h worker, ucp_ep_h ep, 
+                   ucp_ep_ext_proto_t *ep_ext, 
+                   ucp_am_long_hdr_t *hdr, size_t am_length)
+{
+    ucp_am_unfinished_t *parent_am;
+    
+    ucs_list_for_each(parent_am, &ep_ext->am.started_ams, unfinished) {
+        if (parent_am->msg_id == hdr->msg_id) {
+            return parent_am;
+        }
+    }
+    return NULL;
+}
+
+static ucs_status_t
+ucp_am_long_handler_common(void *am_arg, void *am_data, size_t am_length,
+                           unsigned am_flags, ucp_ep_h reply_ep)
+{
+    ucp_worker_h worker = (ucp_worker_h) am_arg;
+    ucp_am_long_hdr_t *long_hdr = (ucp_am_long_hdr_t *) am_data;
+    ucp_ep_h ep = ucp_worker_get_ep_by_ptr(worker, long_hdr->ep);
+    ucp_ep_ext_proto_t *ep_ext = ucp_ep_ext_proto(ep);
+    ucp_recv_desc_t *all_data;
+    uint16_t am_id;
+    size_t left;
+    ucp_am_unfinished_t *unfinished;
+    ucs_status_t status;
+
+    if (ucs_unlikely((long_hdr->am_id >= worker->am_cb_array_len) ||
+		             worker->am_cbs[long_hdr->am_id].cb == NULL)) {
+        ucs_warn("UCP Active Message was received with id : %u, but there" 
+                 "is no registered callback for that id", long_hdr->am_id);
+        goto out;
+    }
+
+    /* if there are multiple messages,
+     * we first check to see if any of the other messages
+     * have arrived. If any messages have arrived,
+     * we copy ourselves into the buffer and leave
+     */
+    unfinished = ucp_am_find_parent(worker, ep, ep_ext, long_hdr, am_length);
+    
+    if (unfinished) {
+        memcpy(UCS_PTR_BYTE_OFFSET(unfinished->all_data + 1, long_hdr->offset),
+               long_hdr + 1, am_length - sizeof(*long_hdr));
+        unfinished->left -= am_length - sizeof(*long_hdr);
+        if (unfinished->left == 0) {
+            am_id = long_hdr->am_id;
+            status = worker->am_cbs[am_id].cb(worker->am_cbs[am_id].context,
+                                              unfinished->all_data + 1,
+                                              long_hdr->total_size,
+                                              reply_ep,
+                                              UCP_CB_PARAM_FLAG_DATA);
+
+            if (status != UCS_INPROGRESS) {
+                ucs_free(unfinished->all_data);
+            }
+
+            ucs_list_del(&unfinished->unfinished);
+            ucs_free(unfinished);
```

Comment:
shouldn't we return here?

Reply:
Yes. Thank you, thats a bad memory leak.